### PR TITLE
perf: use `std::this_thread::sleep_for` instead of busy loop

### PR DESCRIPTION
**Pull request by KowalskiThomas**
## What does this PR do?

This PR changes the sampler loop logic to rely on `std::this_thread::sleep_for` instead of a time-checking busy loop to "plan" for the next sample. 

**Note** I had to update a unit test that was checking the amount of CPU consumed by the sampler thread as it was now not in line with reality anymore. Not sure whether the value I put there is the right one.

**Other note on CI** Retrying jobs allows me to get them greens, I don't know why I have so many reds 🤔  (see https://github.com/KowalskiThomas/echion/pull/3)

## Testing

Benchmarking on simple examples shows significant improvement in CPU use.    
Guidance on how to test/benchmark more widely or in more "real" conditions, please let me know.

### Basic example / proof of concept

Run a few (purely C/C++) threads that do matmul in parallel with a sampler (that does nothing but wake up every once in a while).

```cpp
#include <cstdlib>
#include <ctime>
#include <cstdio>
#include <echion/timing.h>
#include <sched.h>
#include <thread>
#include <vector>

inline unsigned int interval = 1000;

bool get_is_optimized() {
    printf("ECHION_OPTIMIZED: %s\n", getenv("ECHION_OPTIMIZED"));
    const char* env = getenv("ECHION_OPTIMIZED");
    if (env == NULL || atoi(env) == 0) {
        printf("OPTIMIZATION DISABLED\n");
        return false;
    }
    printf("OPTIMIZATION ENABLED\n");
    return atoi(env) == 1;
}

void sampler() {
    last_time = gettime();
    size_t count=0;
    static bool optim = get_is_optimized();
    while (true) {
        count++;
        if (count % 100 == 0) {
            // printf("Sampling...\n");
        }
        auto now = gettime();
        auto end_time = now + interval;

        if (optim) {
            struct timespec ts{
                static_cast<time_t>(end_time / 1000000),
                static_cast<long>(end_time % 1000000)
            };
            nanosleep(&ts, nullptr);
        } else {
            while (gettime() < end_time) {
                sched_yield();
            }
        }

        last_time = now;
    }
}

void matmul() {
    auto a = new int[100][100]();
    auto b = new int[100][100]();
    auto c = new int[100][100]();


    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            for (int k = 0; k < 100; k++) {
                a[i][j] += b[i][k] * c[k][j];
            }
        }
    }

    delete[] a;
    delete[] b;
    delete[] c;
}

void real_function() {
    for (int i = 0; i < 1000; i++) {
        matmul();
    }
    printf("Real function done\n");
}

int main() {
    std::thread sampler_thread(sampler);
    sampler_thread.detach();

    std::vector<std::thread> real_threads;
    for (int i = 0; i < 10; i++) {
        real_threads.push_back(std::thread(real_function));
    }
    for (auto& thread : real_threads) {
        thread.join();
    }
    std::thread real_thread(real_function);

    real_thread.join();
    return 0;
}
```

This was compiled (with `O0`) and executed with `perf`:

```
ECHION_OPTIMIZED=(1|0) sudo -E perf stat -e context-switches,cs,cpu-migrations,syscalls:sys_enter_nanosleep,syscalls:sys_enter_sched_yield,syscalls:sys_enter_clock_nanosleep -- ./a.out
```

The results are pretty compelling:

```
BASE VERSION (not optimized)

Performance counter stats for './a.out':

               145      context-switches                                                                                                               
                16      cpu-migrations                                                        
                 0      syscalls:sys_enter_nanosleep                                          
        16,714,095      syscalls:sys_enter_sched_yield                                        
                 0      syscalls:sys_enter_clock_nanosleep                                      

      13.624349398 seconds time elapsed

      78.082270000 seconds user
       8.238943000 seconds sys

OPTIMIZED VERSION (optimized)

Performance counter stats for './a.out':

           113,808      context-switches                                                                                                           
               204      cpu-migrations                                                        
                 0      syscalls:sys_enter_nanosleep                                          
                 0      syscalls:sys_enter_sched_yield                                        
           113,687      syscalls:sys_enter_clock_nanosleep                                      

      13.404751332 seconds time elapsed

      71.877161000 seconds user
       0.182027000 seconds sys
```

What we see is
- User time is slightly better
- System time is significantly better (>10x)
- Optimised version has way more context switches (which is expected given that it explicitly asks to context switch)
- Optimised version makes way less system calls (one per sampler loop iteration instead of as-many-as-needed per sampler loop iteration)

### In Echion

Compiling Echion and running it (against a simple `matmul` loop  Python file) shows similar results. Counting the number of samples taken yields similar results as well, meaning changing the "scheduling logic" of the sampler does not result in unintended skew/shift in sampling timing.

One noteworthy thing, though, is that at extreme values of `interval` (e.g. `1us` instead of the default `1000us`/`1ms`), the performance degrades (we yield control back to the scheduler whereas using the busy loop would save us some useless system calls) but I don't think taking a sample every microsecond makes a lot of sense as it would mean not leaving much space for the profiled code to run.

✅  Passing `dd-trace-py` PR: https://github.com/DataDog/dd-trace-py/pull/14564

### CODE REVIEW

File: tests/test_cpu_data.py
Comment (P403n1x87) [inline tests/test_cpu_data.py:1]: Why are these changes needed? With an essentially equivalent wait logic I wouldn't have expected changes to test sources 🤔

Comment (KowalskiThomas) [reply tests/test_cpu_data.py:1]: This is illustrated pretty well by this:

```cpp
#include <sched.h>
#include <thread>

#include <time.h>

typedef unsigned long microsecond_t;

#define TS_TO_MICROSECOND(ts) ((ts).tv_sec * 1e6 + (ts).tv_nsec / 1e3)
#define TV_TO_MICROSECOND(tv) ((tv).seconds * 1e6 + (tv).microseconds)

static microsecond_t gettime()
{
    struct timespec ts;
    if (clock_gettime(CLOCK_BOOTTIME, &ts))
        return 0;

    return TS_TO_MICROSECOND(ts);
}


int main()  {
    auto seconds_to_wait = 1;
    
    if (false) { 
        auto end = gettime() + seconds_to_wait*1000000;
        while (gettime() < end) {
            sched_yield();
        } 
    } else {
        std::this_thread::sleep_for(std::chrono::seconds(seconds_to_wait));
    }

    return 0;
}
```

Switching between `false` and `true` gives either

**Using `sleep_for`**

```
./a.out  0.00s user 0.00s system 0% cpu 1.002 total
```

**Using the loop**

```
./a.out  0.50s user 0.51s system 99% cpu 1.001 total
```

**Review by P403n1x87 — APPROVED** [review]
